greenpithumb/moisture_sensor.py

```diff
@@ -1,17 +1,44 @@
 class MoistureSensor(object):
     """Wrapper for DHT11 moisture sensor."""
 
-    def __init__(self, adc, channel):
+    def __init__(self, adc, pi_io, channel, gpio_pins, clock):
         """Creates a new MoistureSensor wrapper.
 
         Args:
             adc: ADC(analog to digital) interface to receive analog signals from
                 moisture sensor.
+            pi_io: Raspberry Pi I/O interface.
             channel: ADC channel to which the moisture sensor is connected.
+            gpio_pins: Tuple of length 2 of the Raspberry Pi GPIO pins that
+                the moisture sensor is connected to.
+            clock: Clock interface.
         """
         self._adc = adc
+        self._pi_io = pi_io
         self._channel = channel
+        self._gpio_pins = gpio_pins
+        self._clock = clock
 
     def moisture(self):
-        """Returns moisture level."""
-        return self._adc.read_adc(self._channel)
+        """Returns the moisture level.
+
+        Takes two readings from the moisture sensor with the GPIO pins powering
+        it alternately turned on and off(one in each state). Returns the average
+        of the two readings.
+        """
+        pin_one = self._gpio_pins[0]
+        pin_two = self._gpio_pins[1]
+
+        self._pi_io.turn_pin_on(pin_one)
+        self._pi_io.turn_pin_off(pin_two)
+        moisture_one = self._adc.read_adc(self._channel)
+
+        self._clock.wait(0.1)
+
+        self._pi_io.turn_pin_on(pin_two)
+        self._pi_io.turn_pin_off(pin_one)
+        moisture_two = 1023 - self._adc.read_adc(self._channel)
+
+        self._pi_io.turn_pin_off(pin_two)
```

Comment:
It might be a good idea to guarantee we always leave both pins in an off state, even if `read_adc` throws an exception (we can use the `finally` keyword for this). Can we add a unit test for this, too?